packages/editor/src/components/post-actions/index.js

```diff
@@ -36,7 +36,10 @@ const POST_ACTIONS_WHILE_EDITING = [
 	'view-post',
 	'view-post-revisions',
 	'rename-post',
+	'rename-template',
```

Comment:
> I'll ramble a bit and might be missing things, and I will check the code and changes better, but it seems too complex to me at first glance.
>
> […]
> 
> I think it's apparent that we need to reuse some actions and know some context (such as 'edit' mode, or 'post vs site' editor for redirections). Can we extract this info somehow to avoid such explicit declarations? If we cannot avoid that, shouldn't it be part of the `actions` API?

I think your outside perspective is very useful, don't ignore your intuition. We need more [red teams](https://en.wiktionary.org/wiki/red_team), in fact.

Taking a step back, there are several things we can do, as well as some questions to ask ourselves now that we have concrete experience with the unification:
* You touched on the idea of context (`edit`, `post`, etc.). You also touched on the idea of a generic action provider potentially conflicting with type-specific requirements: is it a code smell if a function called `isTemplateRevertable` needs to validate its argument as template-like?
* Meanwhile, the current system is very flexible in that each action is a set of methods callable on any data: `isEligible( item )`, `onActionPerformed: (actionId, items) => switch (actionId) {...`. So we are defaulting to giving consumers almost maximum control with the goal of making Data Views versatile and post-type-agnostic.
* Now that we've put things **in**, how much can we take **out**? Could we do the exercise of removing as much of that control as possible, with the goal of minimising moving parts? Then we would be forced to make assumptions about the data and the permissible set of actions. For example, how do we reuse actions across screens and entities (and how much should we)? As another example, what would it take to kill `onActionPerformed` and let the system navigate when necessary or intelligently perform other kinds of post-action cleanup?

Reply:
It seems I had accidentally deleted a small part of my first comment. My main issue is what we are trying to unify.

>Meanwhile, the current system is very flexible in that each action is a set of methods callable on any data: isEligible( item ), onActionPerformed: (actionId, items) => switch (actionId) {.... So we are defaulting to giving consumers almost maximum control with the goal of making Data Views versatile and post-type-agnostic.

That's one of the issues here. This PR (and at least this [merged one](https://github.com/WordPress/gutenberg/pull/60486))  is not about DataViews, but it tries to do what DataViews do.

I worked on the DataViews actions API and you can say it consists of two main parts:
1. The actions API - how we define actions, isEligible, RenderModal, etc..
2. The component that takes as input the above actions and can render them in form of a dialog, inline, etc.. based on the action's properties.

The component that renders the actions is reused from all consumers, but the actions are provided by the consumers. What I'm trying to say is that we should probably have a reusable component to render the actions, but the actions unification is a whole different matter and the unification of them would involve questions like:

>is it to have for example a single delete-post action and apply to every post type and check if isEligible


Right now in trunk and with this PR, we are taking an approach where we duplicate (almost) the [rendering component](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-actions/index.js#L114), which is not necessarily bad to start with, if we don't want the UI to match, but if we want it to match we could consider extracting this to a private component package. Again it's probably fine to just duplicate for now and  update the comments to have in mind if we'll need to create a shared component.

I think we should clarify the vision about actions unification. I don't think it's great to have `renamePostAction, renameTemplateAction` in the same array. It feels weird to me that we already have override `actionIds` in `usePostActions` and `POST_ACTIONS_WHILE_EDITING` to pass. It feels like we don't have clear answers about the use cases and we rush into abstractions, which add complexity. We should either for start have a single rename action etc.. or we could just start by separating the actions per our conditions (probably by checking post type? 🤔 ). If the context matters, we should also check this if/how to incorporate this..

Finally maybe this is the time that we should also just think about a bit, how the actions API is similar to the commands API and if it would be possible to converge/unify a bit there or even take inspiration, or even see gaps in that API 😄 

To sum up, I think we should take things step by step and try to avoid unnecessary complexity.

Reply:
> Finally maybe this is the time that we should also just think about a bit, how the actions API is similar to the commands API and if it would be possible to converge/unify a bit there or even take inspiration, or even see gaps in that API

This reminded me of Riad's [proposal](https://github.com/WordPress/gutenberg/issues/59004): a single API to describe actions/commands and reuse them.

Reply:
Hi @mcsf,
Adding the template check on isTemplateRevertable does not make sense at all 😅 It is something I will change.

> Now that we've put things in, how much can we take out? Could we do the exercise of removing as much of that control as possible, with the goal of minimising moving parts? Then we would be forced to make assumptions about the data and the permissible set of actions. For example, how do we reuse actions across screens and entities (and how much should we)? As another example, what would it take to kill onActionPerformed and let the system navigate when necessary or intelligently perform other kinds of post-action cleanup?

I feel we can not take out onActionPerformed it is very dependent on the context of where the action is executed. E.g: when I delete a template that is assigned to a page I should not navigate anywhere I should just switch to page editing and assign the default template to that page, on data views when a template is deleted I should do nothing at all, and on site editor, I should navigate to the template list. In the view of another plugin that also offers a way to delete templates, I may need to do something totally different. Detecting all these cases and doing the right thing all the time seems "too magic" to me. But I'm open to suggestions of other things we can remove, and or enhancements we could do. The onActionPerformed is part of usePostActions which is private and leaves the door open for enhancements.

Reply:
>The end goal of this code is to provide the same actions on dataviews, edit site inspector, edit site site details panel, and edit post inspector. Which was on mockups. This means that actions like delete template and revert a template should be available on edit post too. So the editor package seemed to be the best package to have the actions.

Probably yes, but are we 100% sure that every DataViews action should be the same? Probably they should.. It's good to have them in editor package and that wasn't the problem for me. 

>Yes in the future we may join the delete-post, delete-template, and delete-pattern action in a single action that can be applied to every post. It leaves the possibility for that and I think it may make sense. But it is not the goal.

I think we should do that and that's the goal. I agree we can do it iteratively, but that means we should separate these actions for start (ex. Template actions, post actions, etc..) and then (if not from that iteration) have shared actions.

For me the complexity stems from the fact that we're trying to infer some info in places but on the same time we have hardcoded strings/actions etc.. I think for start we should:
1. Move the actions to editor package and organise(export) them based on post type probably
2. Let the consumer pass the actions and have a dumb component to render actions